### ISSUE TITLE
Problem with FLOPs

### ISSUE THREAD

**Issue by AlanHuang1998** [google-research/morph-net#105]
Hi, thanks for sharing open source.
Actually I got a problem with FLOPs when I'm training resnet_cifar_10 with PocketFlow.
I try several combinations of regularization_strength and gamma_threshold.
want to ask if this is correct?
![q1](https://user-images.githubusercontent.com/36326341/61277273-6dead780-a7e4-11e9-97ea-cfe34a30f7a9.jpg)

Code for my above implementation is provided below :

`
      
      # data input pipeline
      with tf.variable_scope(self.data_scope):
        iterator = self.build_dataset_train() if is_train else self.build_dataset_eval()
        images, labels = iterator.get_next()
        if not isinstance(images, dict):
          tf.add_to_collection('images_final', images)
        else:
          tf.add_to_collection('images_final', images['image'])
      # model definition - distilled model
      if self.enbl_dst:
        logits_dst = self.helper_dst.calc_logits(sess, images)
      

      # model definition - primary model
      with tf.variable_scope(self.model_scope):
        # forward pass
        if is_train and self.forward_w_labels:
          logits = self.forward_train(images, labels)
        else:
          logits = self.forward_train(images) if is_train else self.forward_eval(images)
        if not isinstance(logits, dict):
          tf.add_to_collection('logits_final', logits)
        else:
          for value in logits.values():
            tf.add_to_collection('logits_final', value)
            
        
        # loss & extra evalution metrics
        loss, metrics = self.calc_loss(labels, logits, self.trainable_vars)
        if self.enbl_dst:
          loss += self.helper_dst.calc_loss(logits, logits_dst)
        tf.summary.scalar('loss', loss)
        for key, value in metrics.items():
          tf.summary.scalar(key, value)

        # optimizer & gradients
        if is_train:
          self.global_step = tf.train.get_or_create_global_step()
          lrn_rate, self.nb_iters_train = self.setup_lrn_rate(self.global_step)
          
          network_regularizer = flop_regularizer.GammaFlopsRegularizer(
            output_boundary = [logits.op],
            input_boundary=[images.op, labels.op],
            gamma_threshold=8.8e-1 # 
          )#
          exporter = structure_exporter.StructureExporter(network_regularizer.op_regularizer_manager)
          regularization_strength = 1e-8 #
          regularizer_loss = (network_regularizer.get_regularization_term() * regularization_strength)#
          model_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits = logits, labels= tf.argmax(tf.cast(labels, dtype=tf.int32),1))#
          optimizer = tf.train.MomentumOptimizer(lrn_rate, FLAGS.momentum)
          train_op = optimizer.minimize(model_loss + regularizer_loss)#
          tf.summary.scalar('RegularizationLoss', regularizer_loss)#
          tf.summary.scalar(network_regularizer.cost_name, network_regularizer.get_cost()) 
          tf.summary.FileWriter(FLAGS.log_dir, graph=tf.get_default_graph())
          exporter = structure_exporter.StructureExporter(network_regularizer.op_regularizer_manager) 
          
          sess.run(tf.global_variables_initializer())
          for step in range(FLAGS.steps):
            _, structure_exporter_tensors = sess.run([train_op, exporter.tensors])
            if (step % 1000 == 0):
              exporter.populate_tensor_values(structure_exporter_tensors)
              exporter.create_file_and_save_alive_counts(FLAGS.log_dir, step)
               `

**Comment by ayp-google** [comment]
One thing that stands out to me is that gamma_threshold is set to 0.88 which is rather large.  The value quoted in the paper is 1e-2.  As a sanity check, you can analyze the checkpoint to generate a histogram of the gamma values as in Figure 2.  That can help choose the right value for gamma_threshold.

**Comment by Huang-Yan-wei** [comment]
Thanks for replying @ayp-google.
As I known, gamma_threshold is a arg for network_regularizer, how I can check my gamma_threshold with histogram?

**Comment by ayp-google** [comment]
You can use something like [inspect_checkpoint.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/inspect_checkpoint.py) to examine the tensor values.  Specifically, you can get all the gamma values and plot a histogram as in Figure 2 in the paper.  There is usually a bimodal distribution and you can choose gamma_threshold to be in the gap.

**Comment by eladeban** [comment]
Note:
It's probably *not*  the problem, but 
` input_boundary=[images, labels]`
should  be:
`input_boundary=[images.op, labels.op]`

**Comment by Huang-Yan-wei** [comment]
Thanks you @eladeban .
I've tried this, but it seems not much different by visualize.
I'll edit it

**Comment by Huang-Yan-wei** [comment]
thanks for your help @ayp-google ,
I tried to plot a histogram of gamma value like Figure 2 in the paper, and I tried to use [inspect_checkpoint.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/inspect_checkpoint.py)
.
Then I got the list like this.
![q2](https://user-images.githubusercontent.com/36326341/61525809-bf48d000-aa4b-11e9-994d-ac8be831dd15.jpg)

I use this to trace on the tensorboard, then I found this.
![q3](https://user-images.githubusercontent.com/36326341/61525951-0931b600-aa4c-11e9-8a8c-5e3381f05dee.jpg)

Want to ask how can I find the gamma value to plot a histogram.

**Comment by ayp-google** [comment]
You can use any [histogram](https://docs.scipy.org/doc/numpy/reference/generated/numpy.histogram.html) tool to plot the histogram.  Basically, you can visualize how the gamma values are distributed and adjust the regularization_strength and gamma_threshold accordingly.

For example, in your tensor all the gamma values are >1.  If so, you may need higher regularization_strength.  Large gamma values indicate the output of this convolution is strong and those channels should not be removed.  Larger regularization strength means gamma value has increased cost so the optimizer will decrease gamma to optimize the overall cost.  Once the gamma value is small, this indicates the output signal is weak and does not contribute much to the prediction so it can be removed.

**Comment by Huang-Yan-wei** [comment]
Thanks for helping @ayp-google.
I add code to my model, code is provided below :
`

          **variables = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, 
                  scope='model/resnet_model/batch_normalization_18/gamma')**
          
          network_regularizer = flop_regularizer.GammaFlopsRegularizer(
            output_boundary = [logits.op],
            input_boundary=[images.op, labels.op],
            gamma_threshold=8.8e-1 # 
          )#
          regularization_strength = 1e-8 #
          regularizer_loss = (network_regularizer.get_regularization_term() * regularization_strength)#
          model_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits = logits, labels= tf.argmax(tf.cast(labels, dtype=tf.int32),1))#
          optimizer = tf.train.MomentumOptimizer(lrn_rate, FLAGS.momentum)
          train_op = optimizer.minimize(model_loss + regularizer_loss)#
          tf.summary.scalar('RegularizationLoss', regularizer_loss)#
          tf.summary.scalar(network_regularizer.cost_name, network_regularizer.get_cost())#FLOPs
          **tf.summary.histogram('gamma', variables)**
          tf.summary.FileWriter(FLAGS.log_dir, graph=tf.get_default_graph())
`

So I got the following histogram, where < 0.88 seems to be deleted, I want to check full gamma histogram without pruning, maybe I should change something?
![q4](https://user-images.githubusercontent.com/36326341/61623361-14cde880-aca9-11e9-9473-841f3544734b.jpg)

**Comment by Huang-Yan-wei** [comment]
@ayp-google , the original gamma value in Figure 1 and gamma with morphnet in Figure 2.
Figure 1
![q5](https://user-images.githubusercontent.com/36326341/61696066-3b525900-ad67-11e9-9fd3-0ce8f274f308.jpg)
Figure 2
![q6](https://user-images.githubusercontent.com/36326341/61696069-3d1c1c80-ad67-11e9-88e7-16e23af5ee4a.gif)

I try to plot the histogram with figure2's value, but it's not work, so I plot  the Fig1's.
![q4](https://user-images.githubusercontent.com/36326341/61697746-7f932880-ad6a-11e9-9bd8-8b569631a842.jpg)


I fine tune gamma_threshold, then I get this, the green one's threshlod is 0.01, gray one is 0.86, red one is 0.95 and my reguralizier strength fixed at 1e-10.
![q7](https://user-images.githubusercontent.com/36326341/61697074-2080e400-ad69-11e9-8857-beda72cd22f9.jpg)

FLOPs still rise up.
I review the paper and get a question, should I use both of flop_reg and modelsize_reg?

**Comment by Cheng-Hsiung** [comment]
Hi, did you check if your total loss (model_loss + regularization_loss) decreases with time?
Also if you train from cifar10 from scratch, the initial softmax classification loss shall be ~ ln10 ~2.3, I suppose? While your regularization starts around 0.017 which is relatively low. 
Hope this helps .

**Comment by eladeban** [comment]
Thanks for sharing this case.
It is very unusual. I don't think I have seen gammas go above 1.0 in any other model. 
I am not sure exactly what model you are using an how are you configuring batch norm parameters but I suspect it could be related to that.

Things that come to mind:

- What is the distribution of gammas when training without the FLOPs regularizer?
- Could you try applying a small weight decay on the gammas (as is done in resnet models, say 1e-4 or so), this could prevent the swell in gammas.
- what is the loss and what is the accuracy of the models you have tried, are they kind of OK? or are they junk?
- The fact that the `regularization_term` is going up is not that surprising as gammas are >1 and the upper bound become more and more loose. The experiments with different thresholds actually suggest to me that you should try exploring the main knob `regularization_strength`, as ayp@ suggests you should try values *larger* than 1e-8,  rather than smaller values. What about 5e-8, 1e-7, 5e-7 with any threshold you see fit.
- you are not saying anything about your model. how non standard is is? are you using a custom LR schedule (i.e. is it increasing, after 3K?) are you using custom initializers?
- I don't recall having problems with Momentum in the past, but could you try RMSPROP with *constant* learning rate of 0.1, just to rule this out? 
- When you say you are using PocketFlow, what do you mean by that, do you have other optimization going on? if so could you try *not* having them on at first.

Hope it helps,

Elad

**Comment by Huang-Yan-wei** [comment]
@Cheng-Hsiung  thanks for sharing :).

I try larger strength and g_threshold fixed at 0.88, the gray one is 1e-10, orange is 1e-2, red one is 1e+2.
![q8](https://user-images.githubusercontent.com/36326341/61764059-6772e600-ae0a-11e9-95ff-700ca2faa555.jpg)

Here is the gamma value, it seems to have decreased in some batch_normalization.
![q9](https://user-images.githubusercontent.com/36326341/61764170-ce909a80-ae0a-11e9-8cdc-16dfc49b2c25.jpg)

I may not notice something, I'll follow Elad's point to improve.

**Comment by Huang-Yan-wei** [comment]
@eladeban 
I have 2 questions. 

- As Fig 1 and Fig 2. Why my json file's conv seems to decreace but my FLOPs still going up. 

- I think it related to weight and bias,as I know it seems to transfer weight, does MorphNet transfer bias?

Fig 1
![q8](https://user-images.githubusercontent.com/36326341/61784390-327f8700-ae3c-11e9-898d-8e005ba8b138.jpg)

FIg 2
![q10](https://user-images.githubusercontent.com/36326341/61784371-2b587900-ae3c-11e9-9db1-04f771f6dc1a.jpg)

Thanks for replying:)

**Comment by AlanHuang1998** [comment]
Everyone, thanks for help :).
I changed my loss and finally did it.
The problem is that I didn't add regularizer_loss to original loss and put regularizer in the wrong place in the code.
